### PR TITLE
Chapter: Architecture

### LINKED ISSUE #9: Chapter: BFFs
Talking about when frontend needs a backend, for proxying, standardizing if they dont have API gateways and shits all over the place, hiding tokens, sorting out cache rules when its all nonsense.

### PR DESCRIPTION
**Pull request by philsturgeon**
Closes #9.

This is a draft, and I need to add some reference links and whatnot. I'd also like to bolster the case for GraphQL beyond "meh IMO its daft" so I'd love your help there GQL fans. 

Also if the REST BFF stuff - or anything else - needs more splainin' please do point out what. But, bear in mind, this is the last chapter in a 12 chapter book, not a blog post. If something feels missing maybe its in a previous chapter. I can reference previously written stuff, but also maybe I forgot to reference it. This is what editors are for but y'all my editors. 😅

### CODE REVIEW

File: 12-architecture.adoc [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+which is using JSON Schema for a type system, or it can wrap gRPC, SOAP, and all sorts of other dependencies, all with one consistent interface.
+
+A REST API operating over HTTP also has the chance to leverage all of HTTP instead of just a subset, providing solid HTTP caching conventions, full HTTP/2 support, and hypermedia controls. 
+
+In my experience the "REST API for Frontend Groups" argument would have solved some rather frustrating problems. An iOS and a Web version of the same internal application was consistently receiving bug reports, where the same employee could see a button on the web app, but could not on the iOS app, even though it should have been the same. 
+
+This was due to some complex if statements checking 5 different things, and one of the apps forgot to check a 6th thing, so the state was out of whack. Hoisting these states up into the REST API mean that one API can interpret arbitrary data from all over the ecosystem. This data could be from hypermedia controls on other REST APIs, or random flags and settings elsewhere, or GraphQL APIs, or a gRPC API, or a local Redis DB, or based on OAtuh scopes, or any combination of the above.
```
Comment (cjunge-work) [inline 12-architecture.adoc]: OAuth scopes

**Review by cjunge-work — CHANGES_REQUESTED** [review]

File: 12-architecture.adoc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+This is not always the case. Google Maps ditched their free mapping offering, Facebook regularly just turn off various APIs for whatever reason, and Meetup decided to ditch their API for any customers who were not Pro. 
+
+Stuff can just vanish, especially if its a third-party API. If you are coupling your client application with a direct dependency on an API, the impact of that API vanishing can be devastating. 
+
+In the past I've worked at companies who used alternative mapping solutions which got bought out by some bigger company and immediately sunset. We had to pull a rush job to switch all mobile apps, web apps, and backend services over to Google Maps, and deploy them all. When we did we made a generic "mapping service" which would proxy requests to Google Maps in the background, meaning we could switch mapping providers more easily, and it end-users on mobile/desktop apps were not stuck with a broken experience until they upgraded.
```
Comment (tstuttard) [inline 12-architecture.adoc]: > we could switch mapping providers more easily, and it end-users on mobile/desktop apps were not stuck with a broken experience until they upgraded.

to 

> we could switch mapping providers more easily, and it **meant** end-users on mobile/desktop apps were not stuck with a broken experience until they upgraded.

File: 12-architecture.adoc [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+== Direct Dependencies
+
+The first thing that comes to mind when trying to interact with a dependency is just interact with that dependency directly. 
+
+Some folks hard-code `api.example.com` into code (don't do this), some folks use configuration managers and environment variables to store the IP/hostname, and some folks use service discovery to automatically update the IP/hostname when it changes for various reasons (like the primary crashing and a secondary kicking in, A/B deployments, etc.)
+
+However a client application is getting the IP/hostname for a dependency, if calls are being made directly to the dependency there are fewer moving parts, but a lot of things that can go wrong. 
+
+_Chapter 9: Timeouts, Retries and Circuit Breakers_ covered a lot of the initial concerns with what can go wrong here. Mainly around remember to set timeouts in all client-side calls. Service mesh helps a lot with handling that stuff at the platform level, without needing to write language-specific code, so if you can funnel your outgoing requests through a service mesh you will get can get a lot of protection against instability, and a lot of insight when things go wrong.
```
Comment (notEthan) [inline 12-architecture.adoc]: > Mainly around remember to set timeouts in all client-side calls.
grammar

File: 12-architecture.adoc [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+A more lightweight debugging/tracing proxy like https://httptoolkit.tech/[HTTP Toolkit] if you need more insight into what is going on if a service mesh is out of reach, but relying on logs alone is probably not going to provide the information you need when a direct dependency is acting up.
+
+=== Change
+
+Hopefully the API is using a reasonable approach to change management: they communicate upcoming change a long way ahead of time, offer alternative solutions to anything being deprecated, and chase applications still using the old approach. 
+
+This is not always the case. Google Maps ditched their free mapping offering, Facebook regularly just turn off various APIs for whatever reason, and Meetup decided to ditch their API for any customers who were not Pro.
```
Comment (notEthan) [inline 12-architecture.adoc]:
```suggestion
This is not always the case. Google Maps ditched their free mapping offering, Facebook regularly just turns off various APIs for whatever reason, and Meetup decided to ditch their API for any customers who were not Pro. 
```

File: 12-architecture.adoc [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+This is not always the case. Google Maps ditched their free mapping offering, Facebook regularly just turn off various APIs for whatever reason, and Meetup decided to ditch their API for any customers who were not Pro. 
+
+Stuff can just vanish, especially if its a third-party API. If you are coupling your client application with a direct dependency on an API, the impact of that API vanishing can be devastating. 
+
+In the past I've worked at companies who used alternative mapping solutions which got bought out by some bigger company and immediately sunset. We had to pull a rush job to switch all mobile apps, web apps, and backend services over to Google Maps, and deploy them all. When we did we made a generic "mapping service" which would proxy requests to Google Maps in the background, meaning we could switch mapping providers more easily, and it end-users on mobile/desktop apps were not stuck with a broken experience until they upgraded.
```
Comment (notEthan) [inline 12-architecture.adoc]:
```suggestion
In the past I've worked at companies who used alternative mapping solutions which got bought out by some bigger company and immediately sunset. We had to pull a rush job to switch all mobile apps, web apps, and backend services over to Google Maps, and deploy them all. When we did we make a generic "mapping service" which would proxy requests to Google Maps in the background, meaning we could switch mapping providers more easily, and end-users on mobile/desktop apps were not stuck with a broken experience until they upgraded.
```

File: 12-architecture.adoc [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+A BFF is an API built for the specific needs of a client. This sounds counterintuitive at first because many of us were taught that APIs must be so generic they can be used by multiple clients, even if those clients require rather different content.
+
+image::images/arch-bff.jpg[]
+
+Trying to make an API so generic that it is useable for every client usually means it is not optimized for any clients. That might be fine depending on how many clients there are, or what they are doing, or how much traffic is coming through the ecosystem, but eventually generic APIs become a source of frustration for the developers who want to build more optimized workflows or data models. 
+
+The worse the design of the remote models the more this starts to become ideal. If you are talking to one of those HTTP/1.1-based RESTish APIs which has jammed 25 resources into a single payload for some reason, but your iPhone application only needs 3 of those resources, making all that extra data go over the wire rather quickly becomes annoying. 
+
+A BFF is a chance for your team to optimize payloads, by grabbing these larger chunks of data and returning only what thinks you probably need.
```
Comment (notEthan) [inline 12-architecture.adoc]: not sure if this is supposed to be 'returning only what it [the BFF] thinks you probably need' or 'returning only what things you probably need'

File: 12-architecture.adoc (hunk 5 above)
Comment (notEthan) [reply 12-architecture.adoc]: removed 'it' from the last sentence

File: 12-architecture.adoc [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+The worse the design of the remote models the more this starts to become ideal. If you are talking to one of those HTTP/1.1-based RESTish APIs which has jammed 25 resources into a single payload for some reason, but your iPhone application only needs 3 of those resources, making all that extra data go over the wire rather quickly becomes annoying. 
+
+A BFF is a chance for your team to optimize payloads, by grabbing these larger chunks of data and returning only what thinks you probably need. 
+
+BFFs also let you solve authentication inconsistencies between the various APIs a client needs to talk to, by letting the end user authenticate with your BFF, then it creates various JWTs, bearer tokens, cookies, and heck knows what else, all stored internally in the BFF API so the client doesn't have to juggle all that nonsense.
```
Comment (notEthan) [inline 12-architecture.adoc]: maybe worth mentioning that often the client should not even have access to direct calls to dependent APIs; the authentication for those may be higher privilege than clients should have.

File: 12-architecture.adoc [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+However a client application is getting the IP/hostname for a dependency, if calls are being made directly to the dependency there are fewer moving parts, but a lot of things that can go wrong. 
+
+_Chapter 9: Timeouts, Retries and Circuit Breakers_ covered a lot of the initial concerns with what can go wrong here. Mainly around remember to set timeouts in all client-side calls. Service mesh helps a lot with handling that stuff at the platform level, without needing to write language-specific code, so if you can funnel your outgoing requests through a service mesh you will get can get a lot of protection against instability, and a lot of insight when things go wrong. 
+
+A more lightweight debugging/tracing proxy like https://httptoolkit.tech/[HTTP Toolkit] if you need more insight into what is going on if a service mesh is out of reach, but relying on logs alone is probably not going to provide the information you need when a direct dependency is acting up.
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply 12-architecture.adoc]: Hey could you do me a solid and use the Suggestion button? It's easier for both of us :D 
image

**Pull request merged**